### PR TITLE
[#5154] refactor(drop-metalake): re-define drop metalake

### PR DESCRIPTION
**Pull request by mchades**
### What changes were proposed in this pull request?

 - Add an in-use property to the metalake with the default value of true.
 - Only empty metalake with in-use=false can be dropped when `force=false`
 - User can use `dorce` option to drop metalake
 - More drop metalake limitations please see the JavaDoc of `dropMetalake` in API module

### Why are the changes needed?

Fix: #5154 

### Does this PR introduce _any_ user-facing change?

yes, users now can not drop an in used metalake

### How was this patch tested?

tests added

### CODE REVIEW

File: clients/client-java/src/main/java/org/apache/gravitino/client/GravitinoAdminClient.java [hunk 1]
```diff
@@ -145,24 +149,74 @@ public GravitinoMetalake alterMetalake(String name, MetalakeChange... changes)
...
   }
 
+  @Override
+  public void enableMetalake(String name) throws NoSuchMetalakeException {
+    ErrorResponse resp =
+        restClient.get(
+            API_METALAKES_IDENTIFIER_PATH + name + "/activate",
+            ErrorResponse.class,
+            Collections.emptyMap(),
+            ErrorHandlers.metalakeErrorHandler());
```
Comment (jerryshao) [inline clients/client-java/src/main/java/org/apache/gravitino/client/GravitinoAdminClient.java:195]: So you didn't update the code here for metalake, it should align with catalog to use `patch` with request body.

File: core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java [hunk 2]
```diff
@@ -104,14 +107,18 @@
...
-      throws NoSuchMetalakeException, NoSuchCatalogException {
-    // todo: check if the metalake is in use
-    return getInUseValue(store, ident);
+  public static void checkCatalogInUse(EntityStore store, NameIdentifier ident)
+      throws NoSuchMetalakeException, NoSuchCatalogException, CatalogNotInUseException,
+          MetalakeNotInUseException {
+    NameIdentifier metalakeIdent = NameIdentifier.of(ident.namespace().levels());
+    checkMetalake(metalakeIdent, store);
+
+    if (!getInUseValue(store, ident)) {
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/catalog/CatalogManager.java:119]: It is better rename to `getCatalogInUseValue` for clear understanding.

File: core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java [hunk 3]
```diff
@@ -116,14 +165,13 @@ public BaseMetalake createMetalake(
       NameIdentifier ident, String comment, Map<String, String> properties)
       throws MetalakeAlreadyExistsException {
     long uid = idGenerator.nextId();
-    StringIdentifier stringId = StringIdentifier.fromId(uid);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java:119]: Why do we remove this?

File: core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java [hunk 4]
```diff
@@ -204,23 +241,112 @@ public BaseMetalake alterMetalake(NameIdentifier ident, MetalakeChange... change
...
+      throws NonEmptyEntityException, MetalakeInUseException {
     try {
-      return store.delete(ident, EntityType.METALAKE);
-    } catch (IOException ioe) {
-      LOG.error("Deleting metalake {} failed due to storage issues", ident, ioe);
-      throw new RuntimeException(ioe);
+      boolean inUse = metalakeInUse(store, ident);
+      if (inUse && !force) {
+        throw new MetalakeInUseException(
+            "Metalake %s is in use, please deactivate it first or use force option", ident);
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java:251]: disable, not deactivate.

File: core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java [hunk 5]
```diff
@@ -204,23 +241,112 @@ public BaseMetalake alterMetalake(NameIdentifier ident, MetalakeChange... change
...
+      if (!catalogEntities.isEmpty() && !force) {
+        throw new NonEmptyEntityException(
+            "Metalake %s has catalogs, please drop them first or use force option", ident);
+      }
+
+      // If reached here, it implies that the metalake is not in use or force is true.
+      if (inUse) {
+        // force is true, so deactivate the metalake first.
+        disableMetalake(ident);
+      }
```
Comment (jerryshao) [inline core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java:265]: The same here for metalake, we don't need to disable this before the drop.

File: server/src/main/java/org/apache/gravitino/server/web/rest/MetalakeOperations.java [hunk 6]
```diff
@@ -149,6 +154,101 @@ public Response loadMetalake(@PathParam("name") String metalakeName) {
...
+      return ExceptionHandlers.handleMetalakeException(OperationType.LOAD, metalakeName, e);
+    }
+  }
+
+  @GET
+  @Path("{name}/activate")
+  @Produces("application/vnd.gravitino.v1+json")
+  @Timed(name = "activate-metalake." + MetricNames.HTTP_PROCESS_DURATION, absolute = true)
+  @ResponseMetered(name = "activate-metalake", absolute = true)
+  public Response activateMetalake(@PathParam("name") String metalakeName) {
```
Comment (jerryshao) [inline server/src/main/java/org/apache/gravitino/server/web/rest/MetalakeOperations.java:199]: Why do we still keep this method and the below one?

File: core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java (hunk 3 above)
Comment (mchades) [reply core/src/main/java/org/apache/gravitino/metalake/MetalakeManager.java:119]: ince the uid was already in the metalake entity field and I don't see that uid being used in properties, I removed it.

File: api/src/main/java/org/apache/gravitino/SupportsMetalakes.java [hunk 7]
```diff
@@ -86,10 +89,69 @@ Metalake alterMetalake(String name, MetalakeChange... changes)
...
    *
-   * @param name The identifier of the metalake.
+   * <ul>
+   *   <li>There is no catalog in the metalake. Otherwise, a {@link NonEmptyEntityException} will be
+   *       thrown.
+   *   <li>The method {@link #disableMetalake(String)} has been called before dropping the metalake.
+   *       Otherwise, a {@link MetalakeInUseException} will be thrown.
+   * </ul>
+   *
+   * It is equivalent to calling {@code dropMetalake(ident, false)}.
```
Comment (jerryshao) [inline api/src/main/java/org/apache/gravitino/SupportsMetalakes.java:101]: `{@code dropMetalake(name, false)}`

**Comment by jerryshao** [comment]
Overall LGTM, just one small comment.

File: api/src/main/java/org/apache/gravitino/SupportsMetalakes.java (hunk 7 above)
Comment (mchades) [reply api/src/main/java/org/apache/gravitino/SupportsMetalakes.java]: fixed

**Review by jerryshao — APPROVED** [review]

**Pull request merged**